### ISSUE TITLE
Wrong / Misleading error message regarding Clone

### ISSUE THREAD

**Issue by Thomblin** [rust-lang/rust#119114]
### Code

```Rust
#[derive(Clone)]
enum Necessity<T> {
    Optional(T),
    Mandatory(T),
}

struct Element<T> {
    attributes: Vec<Necessity<T>>,
    children: Vec<Necessity<Element<T>>>,
}

impl<T: std::clone::Clone> Element<T> {
    fn clone(&self) {
        let _ = self.children.clone();
    }
}

fn main() {
    println!("Hello, world!");
}
```

### Current output

```Shell
Compiling rust_issue v0.1.0 (/home/seeb/rust/rust_issue)
error[E0599]: the method `clone` exists for struct `Vec<Necessity<Element<T>>>`, but its trait bounds were not satisfied
   --> src/main.rs:14:31
    |
2   | enum Necessity<T> {
    | ----------------- doesn't satisfy `Necessity<Element<T>>: Clone`
...
14  |         let _ = self.children.clone();
    |                               ^^^^^ method cannot be called on `Vec<Necessity<Element<T>>>` due to unsatisfied trait bounds
    |
   ::: /home/seeb/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:396:1
    |
396 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    | ------------------------------------------------------------------------------------------------ doesn't satisfy `Vec<Necessity<Element<T>>>: Clone`
    |
    = note: the following trait bounds were not satisfied:
            `Necessity<Element<T>>: Clone`
            which is required by `Vec<Necessity<Element<T>>>: Clone`
help: consider annotating `Necessity<Element<T>>` with `#[derive(Clone)]`
    |
2   + #[derive(Clone)]
3   | enum Necessity<T> {
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rust_issue` (bin "rust_issue") due to previous error
```

### Desired output

```Shell
help: consider annotating `Element<T>` with `#[derive(Clone)]`
    |
2   + #[derive(Clone)]
3   | enum Element<T> {
    |
```

**Comment by BorisBoutillier** [comment]
I think there are two different issue here :
1. rustc effectively points at complicated library internals.
This can be reproduced even with shorter code:
```rust
struct Unclonable;

fn main() {
    let v = vec![Unclonable];
    v.clone()
}
```

Which reports:
```rust
error[E0599]: the method `clone` exists for struct `Vec<Unclonable>`, but its trait bounds were not satisfied
   --> src/main.rs:5:7
    |
1   | struct Unclonable;
    | ----------------- doesn't satisfy `Unclonable: Clone`
...
5   |     v.clone()
    |       ^^^^^ method cannot be called on `Vec<Unclonable>` due to unsatisfied trait bounds
    |
   ::: /home/boris/work/rust/gits/rust/library/alloc/src/vec/mod.rs:398:1
    |
398 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    | ------------------------------------------------------------------------------------------------ doesn't satisfy `Vec<Unclonable>: Clone`
    |
    = note: the following trait bounds were not satisfied:
            `Unclonable: Clone`
            which is required by `Vec<Unclonable>: Clone`
help: consider annotating `Unclonable` with `#[derive(Clone)]`
    |
1   + #[derive(Clone)]
2   | struct Unclonable;
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `issue-119114` (bin "issue-119114") due to 1 previous error
```

2. rustc provides an incorrect suggestion.
In the original code it suggest to add `#[derive(Clone)` to `enum Necessity<T>` while this enum already has the derive, and the real suggestion should be what the author described in the "Desired Output".


Should we create a second issue ?
I think the author main reports was the incorrect suggestion.

**Comment by estebank** [comment]
For the above comment, that output is the current one and is "correct": if you annotate `Unclonable` then the `clone()` call would work.

The misleading output for the original report still points at `Necessity` instead of `Element`.
